cantabular/codebook_test.go

```diff
@@ -0,0 +1,136 @@
+package cantabular_test
+
+import (
+	"bytes"
+	"fmt"
+	"errors"
+	"context"
+	"io/ioutil"
+	"net/http"
+	"testing"
+
+	"github.com/ONSdigital/dp-api-clients-go/cantabular"
+	dphttp "github.com/ONSdigital/dp-net/http"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestGetCodebookUnhappy(t *testing.T) {
+
+	Convey("Given a Cantabular returns an error response from the /Codebook endpoint", t, func() {
+		testCtx := context.Background()
+
+		errorMessage := "this is what cantabular errors look like"
+
+		mockHttpClient := &dphttp.ClienterMock{
+			GetFunc: func(ctx context.Context, url string) (*http.Response, error) {
+				return Response(testErrorResponse(errorMessage), http.StatusNotFound), nil
+			},
+		}
+
+		cantabularClient := cantabular.NewClient(
+			mockHttpClient,
+			cantabular.Config{},
+		)
+
+		Convey("When the GetCodebook method is called", func() {
+			req := cantabular.GetCodebookRequest{}
+			cb, err := cantabularClient.GetCodebook(testCtx, req)
+
+			Convey("Then the expected status code should be recoverable from the error", func() {
+				So(cb, ShouldBeNil)
+				So(cantabular.StatusCode(err), ShouldEqual, http.StatusNotFound)
+			})
+
+			Convey("Then returned error messaage should be extracted correctly", func() {
+				So(err.Error(), ShouldEqual, errorMessage)
+			})
+		})
+	})
+
+	Convey("Given a Cantabular fails to respond", t, func() {
+		testCtx := context.Background()
+
+		mockHttpClient := &dphttp.ClienterMock{
+			GetFunc: func(ctx context.Context, url string) (*http.Response, error) {
+				return nil, errors.New("server failed to respond")
+			},
+		}
+
+		cantabularClient := cantabular.NewClient(
+			mockHttpClient,
+			cantabular.Config{},
+		)
+
+		Convey("When the GetCodebook method is called", func() {
+			req := cantabular.GetCodebookRequest{}
+			cb, err := cantabularClient.GetCodebook(testCtx, req)
+
+			Convey("Then the status code 500 should be recoverable from the error", func() {
```

Comment:
Should this actually be a 502?

Reply:
I think that 500 is OK, as this is just a library talking to Cantabular service. Then the caller of this library should decide what to do. If the caller is an API, then it can return 502 in this scenario, so the caller of the API knows that some service down the line is failing (not the API directly). If the caller is a Kafka service, then it might decide to do something different.